rust_src/remacs-sys/lib.rs

```diff
@@ -85,6 +86,38 @@ pub const INTMASK: EmacsInt = (EMACS_INT_MAX >> (INTTYPEBITS - 1));
 pub const MOST_POSITIVE_FIXNUM: EmacsInt = EMACS_INT_MAX >> INTTYPEBITS;
 pub const MOST_NEGATIVE_FIXNUM: EmacsInt = (-1 - MOST_POSITIVE_FIXNUM);
 
+/// Maximum number of bytes in a buffer.
+/// A buffer cannot contain more bytes than a 1-origin fixnum can
+/// represent, nor can it be so large that C pointer arithmetic stops
+/// working. The ptrdiff_t cast ensures that this is signed, not unsigned.
+//const fn buf_bytes_max() -> ptrdiff_t {
+//    const mpf: ptrdiff_t = (MOST_POSITIVE_FIXNUM - 1) as ptrdiff_t;
+//    const eimv: ptrdiff_t = EmacsInt::max_value() as ptrdiff_t;
+//    const pdmv: ptrdiff_t = libc::ptrdiff_t::max_value();
+//    const arith_max: ptrdiff_t = if eimv <= pdmv {
+//        eimv
+//    } else {
+//        pdmv
+//    };
+//    if mpf as ptrdiff_t <= arith_max {
+//        mpf
+//    } else {
+//        arith_max
+//    }
+//}
+// TODO(db48x): use the implementation above once const functions can have conditionals in them
+const fn buf_bytes_max() -> ptrdiff_t {
+    const p: [ptrdiff_t; 2] = [
+        EmacsInt::max_value() as ptrdiff_t,
+        libc::ptrdiff_t::max_value(),
+    ];
+    const arith_max: ptrdiff_t =
+        p[((p[1] - p[0]) >> ((8 * std::mem::size_of::<ptrdiff_t>()) - 1)) as usize];
+    const q: [ptrdiff_t; 2] = [(MOST_POSITIVE_FIXNUM - 1) as ptrdiff_t, arith_max];
+    q[((q[1] - q[0]) >> ((8 * std::mem::size_of::<ptrdiff_t>()) - 1)) as usize]
+}
+pub const BUF_BYTES_MAX: ptrdiff_t = buf_bytes_max();
+
```

Comment:
Why is this in remacs-sys and not in buffers,rs?

Reply:
Because it's a global variable that we pulled in from the C side of things. It's merely an implementation detail that they wrote it as a macro, and that they put it in buffer.h.

Reply:
Right but why put it in the remacs-sys? At some point we are just going to copy it into buffers.rs so why not just do it now? remacs-sys is "temporary".

Reply:
bad joke :)

When I discovered that it wasn't a real global variable I decided I would move it once I got it working. I actually woke up in the middle of the night with the idea for how to finally implement it in spite of the restrictions on constant functions, so I was probably just tired enough to forget.